internal/gojs/syscall.go

```diff
@@ -163,175 +138,150 @@ func valueSet(ctx context.Context, mod api.Module, sp []uint64) {
-	argsArray := le.Uint32(params[24:])
-	argsLen := le.Uint32(params[32:])
-
-	this := ref(vRef)
-	v := loadValue(ctx, this)
-	m := string(mustRead(mem, "m", mAddr, mLen))
-	args := loadArgs(ctx, mod, argsArray, argsLen)
-
-	var xRef uint64
-	var ok uint32
-	if c, isCall := v.(jsCall); !isCall {
-		panic(fmt.Errorf("TODO: valueCall(v=%v, m=%v, args=%v)", v, m, args))
-	} else if result, err := c.call(ctx, mod, this, m, args...); err != nil {
-		xRef = storeRef(ctx, err)
-		ok = 0
+	var res goos.Ref
+	var ok bool
+	if result, err := c.call(ctx, mod, vRef, m, args...); err != nil {
+		res = storeRef(ctx, err)
 	} else {
-		xRef = storeRef(ctx, result)
-		ok = 1
+		res = storeRef(ctx, result)
+		ok = true
 	}
 
-	// On refresh, start to write results 16 bytes after the last parameter.
-	results := mustRead(mem, "sp", refreshSP(mod)+56, 16)
-
-	// Write the results back to the stack
-	le.PutUint64(results, xRef)
-	le.PutUint32(results[8:], ok)
+	stack.Refresh(mod)
+	stack.SetResultRef(0, res)
+	stack.SetResultBool(1, ok)
 }
 
 // ValueInvoke is stubbed as it isn't used in Go's main source tree.
 //
 // See https://github.com/golang/go/blob/go1.19/src/syscall/js/js.go#L413
-var ValueInvoke = stubFunction(valueInvokeName)
+var ValueInvoke = goarch.StubFunction(valueInvokeName)
 
 // ValueNew implements js.valueNew, which is used to call a js.Value, e.g.
 // `array.New(2)`.
 //
 // See https://github.com/golang/go/blob/go1.19/src/syscall/js/js.go#L432
 // and https://github.com/golang/go/blob/go1.19/misc/wasm/wasm_exec.js#L380-L391
-var ValueNew = newSPFunc(valueNewName, valueNew)
+var ValueNew = goos.NewFunc(valueNewName, valueNew,
+	[]string{"v", "args", "args_len", "padding"},
+	[]string{"res", "ok"},
+)
 
-func valueNew(ctx context.Context, mod api.Module, sp []uint64) {
+func valueNew(ctx context.Context, mod api.Module, stack goos.Stack) {
 	mem := mod.Memory()
+	vRef := stack.ParamRef(0)
+	args := stack.ParamVals(ctx, mem, 1 /*, 2 */, loadValue)
+	// 3 = padding
 
-	// Read param count * 8 memory starting at SP+8
-	params := mustRead(mem, "sp", uint32(sp[0]+8), 24)
-
-	vRef := le.Uint64(params)
-	argsArray := le.Uint32(params[8:])
-	argsLen := le.Uint32(params[16:])
-
-	args := loadArgs(ctx, mod, argsArray, argsLen)
-	ref := ref(vRef)
-	v := loadValue(ctx, ref)
-
-	var xRef uint64
-	var ok uint32
-	switch ref {
+	var res goos.Ref
+	var ok bool
+	switch vRef {
 	case refArrayConstructor:
 		result := &objectArray{}
-		xRef = storeRef(ctx, result)
-		ok = 1
+		res = storeRef(ctx, result)
+		ok = true
 	case refUint8ArrayConstructor:
-		var result *byteArray
-		if n, ok := args[0].(float64); ok {
+		var result interface{}
+		a := args[0]
+		if n, ok := a.(float64); ok {
 			result = &byteArray{make([]byte, uint32(n))}
-		} else if n, ok := args[0].(uint32); ok {
-			result = &byteArray{make([]byte, n)}
-		} else if b, ok := args[0].(*byteArray); ok {
-			// In case of below, in HTTP, return the same ref
+		} else if _, ok := a.(*byteArray); ok {
```

Comment:
fyi I tried to get args to be able to unwrap byteArray, but failed a little, so Ieft it. In any case that can be done in a follow-up PR. Meanwhile I simplified the logic here a little bit and made the reference increment more explicit than before.